### ISSUE TITLE
Virtual thermostat

### ISSUE THREAD

**Issue by villamann** [jvmahon/homebridge-homeseer#63]
Hi, and thanks for development previous help! This time I would like to ask for support of virtual thermostats from a Homeseer plugin called [SDJ-VStat](https://shop.homeseer.com/products/steve-jones-sdj-vstat-software-plug-in-for-hs3). It let you build a thermostat from different kind of sensors and switches.
Unfortuantly it does not go so well with the way thermostats is build with available config parameters in homegridge/homekit. F.ex. SDJ-VStat creates a heating thermostat and/or a cooling thermostat, and the respective mode switch let you choose home/away and not cooling/heating.

Are there any other homebridge parameters that could be involved to make these plugin-thermostats function better? Attached is the user manual for the SDJ-VStat plugin.
[SDJ-VStat Guide.pdf](https://github.com/jvmahon/homebridge-homeseer/files/2532020/SDJ-VStat.Guide.pdf)

**Comment by jvmahon** [comment]
I took a look at the manual and, unfortunately, the way SDJ-VState sets up its devices, it doesn't seem that compatible with how HomeKit handles thermostats.  The only way I can think of to get this to work would be (maybe) if you could set up a bunch of virtual devices that "looks" like a normal thermostat with setpoints, a temperature sensor, a target temperature, etc. and use events to map between the SDJ-Vstat devices and the more typical looking ones. My feeling is this would be a fairly complex undertaking. Other than that suggestion, I don't think I can help on this one.

**Comment by villamann** [comment]
Hi again,

I recently looked into this topic once more taking up your suggestion using virtual devices, and found a quite easy setup that (almost) works.
Using the SDJ-VStat plugin I have an electric oven which is switched on/off with a smart plug in the outlet, controlled by an external temp.sensor. I set the ext. temp sensor as _Ref_ and the Set temp target device in the SDJ-VStat as the _heatingSetpointRef_. For _stateRef/controlRef_ I have created a virtual device that just keeps it at Heating, as that is all I need.
All this works beautiful, until the Temp target is set with a decimal in the Home app. Then nothing changes in HomeSeer and it return an error in the Homebridge Log stating:
`[HomeSeer] Warning - Error Received on Telnet port. If this occurred during startup, this maybe due to bad password / login information in your config.json file. Received data: error, bad format`
Every other switch/device works as it should from the Home app, just sending a value with decimals give the error. Temp target without decimals works fine.
Living in Norway, we use , (comma) as the decimal sign, I suspect that has something to do with it. I have tried switching to US format on both phone and Raspberry, without any luck. Any ideas how to come around this would be highly appreciated!

I use your new Homebridge 4 plugin, but since we had talked about this previously I followed up on this thread.

**Comment by jvmahon** [comment]
I suspect you are right and it is the comma that is causing the problem. On the Telnet interface, a comma separates different data items so if the comma is being sent, that means the telnet interface may think the temperature data is two separate items. I'm not sure if there is a practical way for me to fix this right now - the Thermostat code is particularly tricky - if I adjust it, I have to spend a lot of time testing to make sure nothing else breaks. Unfortunately, as my work life is quite busy right now, I don't have time to do it right, and I don't want to risk doing it wrong, so I won't be able to do anything right now. Sorry about that.

**Comment by villamann** [comment]
No worries at all, fully understandable. However the Thermostat UI in the Home app is so nice compared to my workaround with several switches, so I will still use it and try to avoid half degrees.

If possible I would like to hear your thoughts of where I should focus my effort to try to fix the difference in locale. I can try to change the locale setup in the raspberry, I can also change the signs/locale in the pc running Homeseer, or would you guess it`s the phone enabling the change which causes it?
In other words; which part(s) of the communication chain do you think is the problem?
